chimg/image_customizer.py

```diff
@@ -0,0 +1,241 @@
+        self.target_mount_point = Path(target_mount_point)
+        self.chimg_config_file = Path(chimg_config_file)
+        self.overwrite = overwrite
+
+        self.modified_image_file = self.input_image_file.with_suffix(".modifying")
+        self.output_files_name = self.output_image_path.with_suffix("")
+        self.input_image_type = ""
+        self.resolv_conf_existed = False
+
+        # Check if the output image file already exists
+        if os.path.exists(self.output_image_path) and not self.overwrite:
+            raise FileExistsError(
+                f"Error: Output image file '{self.output_image_path}' already exists! Use overwrite=True to overwrite."
+            )
+
+    def _ensure_paths(self):
+        # if the input image is just in the current directory, we dont need to create the directory
+        if os.path.dirname(self.input_image_file) != "":
+            os.makedirs(os.path.dirname(self.modified_image_file), exist_ok=True)
+        os.makedirs(self.target_mount_point, exist_ok=True)
+
+    def _remove_existing_modified_image(self):
+        if os.path.exists(self.modified_image_file):
+            os.remove(self.modified_image_file)
+            logger.info("Removed existing modified image file")
+
+    def _validate_input_image_exists(self):
+        if not os.path.isfile(self.input_image_file):
+            raise FileNotFoundError(f"Error: Input image file '{self.input_image_file}' not found!")
+
+    def _convert_or_copy_image(self):
+        file_info = subprocess.check_output(["file", self.input_image_file]).decode()
+        if "QCOW" in file_info:
+            logger.info("Converting input qcow2 image to raw image...")
+            subprocess.run(
+                ["qemu-img", "convert", "-f", "qcow2", "-O", "raw", self.input_image_file, self.modified_image_file],
+                check=True,
+            )
+            self.input_image_type = "qcow2"
+        else:
+            logger.info("Image is already raw. Copying to %s", self.modified_image_file)
+            shutil.copy(self.input_image_file, self.modified_image_file)
+            self.input_image_type = "raw"
+        logger.info("Converted or copied image for modifying to: %s", self.modified_image_file)
+
+    def _get_partition_offset(self):
+        fdisk_output = subprocess.check_output(["fdisk", "-l", self.modified_image_file]).decode()
+        start_sector = next((line.split()[1] for line in fdisk_output.splitlines() if "Linux filesystem" in line), None)
+        sector_size_line = next((line for line in fdisk_output.splitlines() if "Sector size" in line), None)
+        sector_size = sector_size_line.split()[3] if sector_size_line else None
+
+        if not start_sector or not sector_size:
+            raise RuntimeError("Error: Could not determine partition offset!")
+
+        offset = int(start_sector) * int(sector_size)
+        return offset
+
+    def _mount_image(self, offset):
```

Comment:
* maybe `losetup` is easier?
* can we use contextmanagers for everything that needs cleanup steps at the end pleaes?